### PR TITLE
Memory access response rework

### PR DESCRIPTION
**Pull request by khauersp**
Reworking of memory access protocol

### CODE REVIEW

**Comment by drewr95** [comment]
I don't want to be the only person reviewing this.  We should get Todd to review this as well

File: j1939/Dm14Query.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    PROCEED = 0
+    BUSY = 1
+    OPERATION_COMPLETE = 4
+    OPERATION_FAILED = 5
+
+
+class Dm14Query:
+    def __init__(self, ca: j1939.ControllerApplication) -> None:
+        """
+        performs memory access queries using DM14-DM18 messaging.  Presently only read queries are supported
```
Comment (snideto) [inline j1939/Dm14Query.py]: This is the stuff that already existed, right?  If so, both read and write queries are, or at least were, supported.

Comment (khauersp) [reply j1939/Dm14Query.py]: Yeah this stuff existed already I broke it out into a new module just to reduce the size of the memory_access module. Read and write queries were already supported but responding to them wasn't

File: j1939/Dm14Server.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        if pgn != j1939.ParameterGroupNumber.PGN.DM14:
+            return
+        if (
+            (self.sa is not None and sa != self.sa)
+            or (
+                self.address is not None and self.address != data[2 : (self.length - 2)]
+            )
+            or self._busy
+        ):
+            sa_temp = self.sa
```
Comment (snideto) [inline j1939/Dm14Server.py]: Having to store all these variables in temp variables seems a little hacky.  Maybe you should just make `_send_dm15()` take args instead?

File: j1939/Dm14Server.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        :param int sa: source address of the message
+        :param int timestamp: timestamp of the message
+        :param bytearray data: data of the PDU
+        """
+
+        if pgn != j1939.ParameterGroupNumber.PGN.DM16 or sa != self.sa:
+            return
+
+        length = min(data[0], len(data) - 1)
+        # assert object_count == self.object_count
```
Comment (snideto) [inline j1939/Dm14Server.py]: uncomment or remove

File: j1939/Dm14Server.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        self.mem_data = data[1 : length + 1]
+        self.data_queue.put(data)
+        self._ca.unsubscribe(self._parse_dm16)
+        self._ca.subscribe(self.parse_dm14)
+        self.state = ResponseState.SEND_OPERATION_COMPLETE
+        self._send_dm15()
+
+    def bytes_to_int(self, data: bytearray) -> int:
+        """
+        Convert bytesaray to integer
```
Comment (snideto) [inline j1939/Dm14Server.py]:
```suggestion
        Convert bytearray to integer
```

File: j1939/Dm14Server.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        self.status = (
+            j1939.Dm15Status.PROCEED.value
+            if proceed
+            else j1939.Dm15Status.OPERATION_FAILED.value
+        )
+        if self.status == j1939.Dm15Status.PROCEED.value:
+            self.state = ResponseState.SEND_PROCEED
+        else:
+            self.state = ResponseState.SEND_ERROR
+        # self._ca.unsubscribe(self._parse_dm14)
```
Comment (snideto) [inline j1939/Dm14Server.py]: Uncomment or remove.

File: j1939/Dm14Server.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        self._wait_for_data()
+        mem_data = None
+        if self.state == ResponseState.WAIT_FOR_DM16:
+            mem_data = self.data_queue.get(block=True, timeout=3)
+
+        return self.mem_data if self.mem_data is not None else None
```
Comment (snideto) [inline j1939/Dm14Server.py]: Does the if/else statement here do anything?

File: j1939/controller_application.py
Comment (snideto) [inline j1939/controller_application.py:1]: Undo the formatting changes in this file.  Only run black on the memory access or other new files.

File: j1939/electronic_control_unit.py
Comment (snideto) [inline j1939/electronic_control_unit.py:1]: Undo the formatting changes.

**Review by snideto — CHANGES_REQUESTED** [review]
Fix "only read supported" comment, uncomment or remove commented code, and undo formatting changes to controller_application and electronic_control_unit.

File: j1939/Dm14Server.py (hunk 6 above)
Comment (khauersp) [reply j1939/Dm14Server.py]: No, not really I'll eliminate it

**Review by drewr95 — APPROVED** [review]

**Pull request merged**